### PR TITLE
Proxy: Filters API

### PR DESCRIPTION
**Pull request by kubistika**
This PR adds initial support for external loadable modules to filter protocol messages. For more information, please read the docs at `proxy/filters/README.md`.

On the way, it fixes: 
* Mem leak in `pf_server_parse_target_from_routing_token` introduced during the previous PR (#5372).
* Segfault when  `pf_server_parse_target_from_routing_token` fails.

### CODE REVIEW

**Comment by akallabeth** [comment]
@freerdp-bot test

**Comment by akallabeth** [comment]
@freerdp-bot test

**Comment by kubistika** [comment]
@akallabeth In this PR we introduced a new feature that supports loading external modules that filters RDP protocol messages. We wanted the modules to be 100% seperated from the proxy itself so we used `inttypes.h` instead of compiling with `winpr` library, so msbuild2010 fails... what do you think we should do to solve it?

**Comment by akallabeth** [comment]
@kubistika that is what we introduced `winpr` for :/

File: server/proxy/pf_server.c [hunk 1]
```diff
@@ -155,12 +161,13 @@ static BOOL pf_server_post_connect(freerdp_peer* client)
...
 	/* keep both sides of the connection in pdata */
 	pc->pdata = ps->pdata;
-	pdata->pc = (pClientContext*) pc;
+	pdata->info->TargetHostname = _strdup(host);
+	pdata->info->Username = _strdup(client->settings->Username);
+	pdata->pc = pc;
 	pdata->ps = ps;
-	pdata->connectionClosed = connectionClosedEvent;
+	pdata->connectionClosed = CreateEvent(NULL, TRUE, FALSE, NULL);
 	pf_server_rdpgfx_init(ps);
```
Comment (hardening) [inline server/proxy/pf_server.c]: missing check

File: server/proxy/pf_server.c [hunk 2]
```diff
@@ -206,7 +213,15 @@ static DWORD WINAPI pf_server_handle_client(LPVOID arg)
...
-	pdata = calloc(1, sizeof(proxyData));
+	pdata = pf_context_proxy_data_new();
+
+	if (pdata == NULL)
+	{
+		WLog_ERR(TAG, "pf_context_proxy_data_new failed!");
+		return 0;
+	}
+
+	pdata->info->ClientHostname = _strdup(client->hostname);
```
Comment (hardening) [inline server/proxy/pf_server.c]: missing check

File: server/proxy/pf_filters.c [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+}
+
+BOOL pf_filters_register_new(filters_list* list, const char* module_path, const char* filter_name)
+{
+	proxyEvents* events = NULL;
+	proxyFilter* filter = NULL;
+	HMODULE handle = NULL;
+	filterInitFn fn;
+
+	if (list == NULL)
```
Comment (hardening) [inline server/proxy/pf_filters.c]: I don't see what could lead to have `list` being `NULL`

**Comment by kubistika** [comment]
@hardening thanks for the comments, i'm working on resolving them and fixing some other issues (such mem checks/leaks)
will comment here when i'm done

**Comment by kubistika** [comment]
@hardening fixed your comments and refactored some of the code. now filters_api.h uses winpr, so it should compile on vs2010 too

File: server/proxy/pf_filters.c (hunk 3 above)
Comment (kubistika) [reply server/proxy/pf_filters.c]: changed that check to assert
might happen if a developer calls this function with filters_list uninitialized

**Comment by akallabeth** [comment]
@freerdp-bot test

**Comment by kubistika** [comment]
@akallabeth vs2010...
can you test again please? there were a lot of errors, might be more but it will be easier for me to fix them after those fixes

**Comment by akallabeth** [comment]
@freerdp-bot test

**Comment by kubistika** [comment]
@akallabeth should be fine now i hope

**Comment by akallabeth** [comment]
@freerdp-bot test

**Comment by akallabeth** [comment]
@kubistika let us see what the oracle has to tell ^^

**Comment by kubistika** [comment]
@akallabeth :( maybe now? fixed the macro with do{ ...} while(0)
btw, i'll be glad if you'll check the disp pr (#5414)

**Comment by akallabeth** [comment]
@freerdp-bot test

File: server/proxy/filters/filter_demo.c [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+	WINPR_UNUSED(event_data);
+
+	return FILTER_PASS;
+}
+
+static PF_FILTER_RESULT demo_filter_mouse_event(connectionInfo* info, void* param)
+{
+ 	proxyMouseEventInfo* event_data = (proxyMouseEventInfo*) param;
+
+    if (event_data->x % 100 == 0)
```
Comment (speidy) [inline server/proxy/filters/filter_demo.c]: indent

**Comment by akallabeth** [comment]
@freerdp-bot test

**Comment by speidy** [comment]
LGTM. lets merge

**Pull request merged**